File: LeetCodeJava/NumberOfIslands_200.java
public class NumberOfIslands_200 {
    public int numIslands(char[][] grid) {
        int m = grid.length;
        int n = grid[0].length;
        int numIslands = 0;

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (grid[i][j] == '1') {
                    numIslands++;
                    dfs(grid, i, j);
                }
            }
        }

        return numIslands;
    }

    private void dfs(char[][] grid, int i, int j) {
        int m = grid.length;
        int n = grid[0].length;
        
        // Check boundaries and if the cell is already visited (water).
        if (i < 0 || i >= m || j < 0 || j >= n || grid[i][j] != '1') {
            return;
        }

        // Mark the cell as visited by changing '1' to '0'.
        grid[i][j] = '0';

        // Visit all four adjacent cells (right, down, left, up).
        dfs(grid, i, j + 1);
        dfs(grid, i + 1, j);
        dfs(grid, i, j - 1);
        dfs(grid, i - 1, j);
    }
}

File: LeetCodeCpp/problems/graphs/NumberOfIslands_200.cpp
#include "NumberOfIslands_200.h"

#include <vector>

using namespace std;

int NumberOfIslands_200::numIslands(vector<vector<char>>& grid)
{
    if (grid.empty() || grid[0].empty()) {
        return 0;
    }

    const int m = static_cast<int>(grid.size());
    const int n = static_cast<int>(grid[0].size());
    int numIslands = 0;

    for (int i = 0; i < m; ++i) {
        for (int j = 0; j < n; ++j) {
            if (grid[i][j] == '1') {
                ++numIslands;
                dfs(grid, i, j);
            }
        }
    }

    return numIslands;
}

void NumberOfIslands_200::dfs(vector<vector<char>>& grid, int i, int j)
{
    const int m = static_cast<int>(grid.size());
    const int n = static_cast<int>(grid[0].size());

    if (i < 0 || i >= m || j < 0 || j >= n || grid[i][j] != '1') {
        return;
    }

    grid[i][j] = '0'; // Mark visited in-place.

    dfs(grid, i, j + 1);
    dfs(grid, i + 1, j);
    dfs(grid, i, j - 1);
    dfs(grid, i - 1, j);
}
